experimental/ActionPolyRing/src/Types.jl

```diff
@@ -0,0 +1,203 @@
+  upoly_ring::AbstractAlgebra.Generic.UniversalPolyRing{T}
+  elementary_symbols::Vector{Symbol}
+  ndiffs::Int
+  are_perms_up_to_date::Bool
+  jet_to_var::Any #Always of type Dict{Tuple{Int, Vector{Int}}, DifferentialPolyRingElem{T}}
+  var_to_jet::Any #Always of type Dict{DifferentialPolyRingElem{T}, Tuple{Int, Vector{Int}}}
+  jet_to_upoly_idx::Dict{Tuple{Int, Vector{Int}}, Int}
+  ranking::Any #Alyways of type DifferentialRanking{T}
+  permutation::Vector{Int}
+
+  function DifferentialPolyRing{T}(R::Ring, nelementary_symbols::Int, ndiffs::Int) where {T}
+    @req nelementary_symbols >= 1 "The number of elementary symbols must be positive"
+    @req ndiffs >= 1 "The number of derivations must be positive"
+    elementary_symbols = map(x -> Symbol("u" * string(x)), 1:nelementary_symbols)
+    upoly_ring = universal_polynomial_ring(R; cached = false)
+    
+    jet_to_var = Dict{Tuple{Int, Vector{Int}}, DifferentialPolyRingElem{T}}()
+    var_to_jet = Dict{DifferentialPolyRingElem{T}, Tuple{Int, Vector{Int}}}()
+    jet_to_upoly_idx = Dict{Tuple{Int, Vector{Int}}, Int}()
+    
+    return new{T}(upoly_ring, elementary_symbols, ndiffs, false, jet_to_var, var_to_jet, jet_to_upoly_idx)
+  end
+ 
+  function DifferentialPolyRing{T}(R::Ring, elementary_symbols::Vector{Symbol}, ndiffs::Int) where {T}
+    @req !is_empty(elementary_symbols) "The number of elementary symbols must be positive"
+    @req ndiffs >= 1 "The number of endomorphisms must be nonnegative"
+    upoly_ring = universal_polynomial_ring(R; cached = false)
+    
+    jet_to_var = Dict{Tuple{Int, Vector{Int}}, DifferentialPolyRingElem{T}}()
+    var_to_jet = Dict{DifferentialPolyRingElem{T}, Tuple{Int, Vector{Int}}}()
+    jet_to_upoly_idx = Dict{Tuple{Int, Vector{Int}}, Int}()
+    
+    return new{T}(upoly_ring, elementary_symbols, ndiffs, false, jet_to_var, var_to_jet, jet_to_upoly_idx)
+  end
+
+end
+
+mutable struct DifferentialPolyRingElem{T} <: ActionPolyRingElem{T}
+  p::AbstractAlgebra.Generic.UniversalPolyRingElem{T}
+  parent::DifferentialPolyRing{T}
+  initial::Any #Always of type DifferentialPolyRingElem{T}
+  permutation::Vector{Int}
+
+  DifferentialPolyRingElem{T}(dpr::DifferentialPolyRing{T}) where {T} = new{T}(zero(dpr.upoly_ring), dpr)
+
+  function DifferentialPolyRingElem{T}(dpr::DifferentialPolyRing{T}, upre::AbstractAlgebra.Generic.UniversalPolyRingElem{T}) where {T}
+    @req dpr.upoly_ring === parent(upre) "The parent does not match"
+    new{T}(upre, dpr)
+  end
+
+  function DifferentialPolyRingElem{T}(dpr::DifferentialPolyRing{T}, mpre::MPolyRingElem{T}) where {T}
+    @req dpr.upoly_ring.mpoly_ring === parent(mpre) "The parent does not match"
+    new{T}(mpre, dpr)
+  end
+
+end
+
+struct DifferentialPolyCoeffs{T <: RingElem}
+   poly::T
+end
+
+struct DifferentialPolyExponentVectors{T <: RingElem}
+   poly::T
+end
+
+struct DifferentialPolyTerms{T <: RingElem}
+   poly::T
+end
+
+struct DifferentialPolyMonomials{T <: RingElem}
+   poly::T
+end
```

Comment:
these could easily be merged with the Difference* variants in the same way as the ranking types, by adding the ActionPolyRing type as a type parameter, reducing some code duplication in the functionality of these.